<!DOCTYPE html>
<html>
<head>
   <title>gkno | Next-Gen Sequencing Software</title>
   <link href='http://fonts.googleapis.com/css?family=Ubuntu:400,700&v2' rel='stylesheet' type='text/css'>
   <link href='css/site.css' rel='stylesheet' type='text/css'>
   <link href='css/dropdown.css' rel='stylesheet' type='text/css'>

   <script type="text/javascript">

     var _gaq = _gaq || [];
     _gaq.push(['_setAccount', 'UA-32956625-1']);
     _gaq.push(['_trackPageview']);

     (function() {
       var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
       ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
       var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
     })();

   </script>      
</head>

<body>
   <img src="images/ribbon.png" class="ribbon"/>      
   <div id="header">
      <img src="images/gkno_logo_medium_me.png" height='104%'/>
      <nav id="nav">
          <ul id="navigation" class="box-shadow">
              <li><a href="/" class="first">Home</a></li>
              <li><a href="#">Software <span style="font-size:0.65em">&#9660;</span></a>
                  <ul>
                      <li><a href="tools.html">Tools</a></li>
                      <li><a href="pipelines.html">Pipes</a></li>
                      <li><a href="#">Data</a></li>
                      <li><a href="https://github.com/gkno/gkno_launcher/wiki/Documentation#wiki-installing">Download</a></li>
                  </ul>
              </li>
              <li><a href="tutorials.html">Tutorials</a></li>
              <li><a href="#">Support <span style="font-size:0.65em;">&#9660;</span></a>

                  <ul>
                      <li><a href="forum.html">Forum</a></li>
                      <li><a href="faq.html">FAQ</a></li>
                      <li><a href="bugs.html">Bugs</a></li>
                      <li><a href="https://github.com/gkno/gkno_launcher/wiki/Documentation">Docs</a></li>
                      <li><a href="glossary.html">Glossary</a></li>
                  </ul>                     
               </li>
              <li><a class="last" href="http://blog.gkno.me">Blog</a></li>
          </ul>
      </nav>     
   
      <div style="clear:both"></div>
   </div>

   <div id="intro" class="box-shadow">
       <div class="grid">
          <div style="padding:20px 30px 20px 30px">
             <div style="font-size: 3em; color: rgb(60,60,60)">Glossary</div>
             <div style="font-size: 2em; color: rgb(60,60,60)">Tools</div>
             <div class="description" style="line-height: 25px">
                <ul>
                   <li>bamtools - <em>alignment manipulation and analysis software.</em></li>
                   <li>bamleftalign - <em>Tool for left aligning indels in primary alignments.</em></li>
                   <li>freebayes - <em>Bayesian variant calling software.</em></li>
                   <li>gatk-lite - <em>Third-party genome analysis software</em></li>
                   <li>mosaik - <em>Hash-based Smith-Waterman alignment software.</em></li>
                   <li>ogap - <em>Gap opening realigner that operates on streaming bam files.</em></li>
                   <li>picard - <em>Third-party genome analysis software</em></li>
                   <li>premo - <em>Application for determining suitable alignment parameters for Mosaik.</em></li>
                   <li>samtools - <em>Third party software for sam/bam file manipulation/analysis.</em></li>
               </ul>
             <div style="font-size: 2em; color: rgb(60,60,60)">File formats</div>
             <div class="description" style="line-height: 25px">
                <ul>
                   <li>ann - <em>File generated by a neural net and used by the Mosaik alignment software for calculating mapping qualities.</em></li>
                   <li>bam - <em>Binary version of <b>sam</b> format.</em></li>
                   <li>bai - <em>Index allowing jumping to specified locations in the <b>bam</b> file.</em></li>
                   <li>dat - <em>Reference genome in a Mosaik native format.  This is a file generated from the <b>fasta</b> file and used within <b>Mosaik</b> for aligning reads against.</em></li>
                   <li>fasta (fa) - <em>File format for reference genomes.  See <a href="http://en.wikipedia.org/wiki/FASTA_format">spec</a> for details.</em></li>
                   <li>fai - <em>Reference genome index</em></li>
                   <li>fastq (fq) - <em>File format for sequencing reads and associated base qualities.  See <a href="http://nar.oxfordjournals.org/content/38/6/1767">spec</a> for details.</em></li>
                   <li>mkb - <em>Mosaik read archive.  Mosaik used a preprocessing step to convert <b>fastq</b> files into this format and then uses these files in the alignment process.</em></li>
                   <li>sam - <em>Sequence alignment/map format.  A format for storing the genomic sequence, position, base and mapping qualities and additional information for sequencing reads.  See <a href="http://samtools.sourceforge.net/SAM1.pdf">spec</a> for details.</em></li>
                   <li>vcf - <em>Variant call format.  A file format for storing genomic variants by genome coordinate.  As well as allele indormation, file can also contain genotype information for any number of samples.  See <a href="http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41">spec</a> for details.</em></li>
               </ul>
             </div>
             <div style="font-size: 2em; color: rgb(60,60,60)">Variant types</div>
             <div class="description" style="line-height: 25px">
                <ul>
                   <li>complex - <em>a complex allele consisting of multiple phased individual variants.</em></li>
                   <li>snp - <em>single nucleotide polymorphism.  A single base allele that differs from the reference allele at that locus.</em></li>
                   <li>indel - <em>an insertion or deletion polymorphism.</em></li>
                   <li>mei - <em>mobile element insertion.</em></li>
                   <li>mnp - <em>multiple nucleotide polymorphism.  A sequence of contiguous nucleotides that differs from the reference allele at that locus.</em></li>
                   <li>sv - <em>structural variation.  Any large scale (>100bp) structural change in the genome, including, but not limited to insertions, deletions, inversions and translocations.</em></li>
               </ul>
             </div>
             <div style="font-size: 2em; color: rgb(60,60,60)">General terms</div>
             <div class="description" style="line-height: 25px">
                <ul>
                   <li>aligning - <em>The process of finding the best position for a sequencing read in a reference genome.  After the best position has been found, the read is said to be aligned.</em></li>
                   <li>phasing - <em>Two alleles are phased if they appear on the same chromosome.</em></li>
                   <li>mapping - <em>See <b>aligning</b>.</em></li>
               </ul>
             </div>
          </div>

       </div>
    </div>

</body>
</html>
