---
layout: page
permalink: cshl15.html
---

<!-- -->
<div class='section'>
  <div class='section-title'>Advanced Sequencing Technologies and Applications: BAM tutorial</div>
  <div class='description'>
    The goal of this page is to highlight a number of tools that can be used to view and manipulate sequence alignment (BAM) files. These tools are all included in the gkno package, so no additional tools need to be downloaded to work through these exercises.
  </div>

  <!-- Preparation -->
  <div class='section-subtitle'>Set up</div>
  <div class='description'>
    To begin, create a new directory call 'bam-demo' and copy one of the <i>gkno</i> resource files across to this directory. We will also use some bash magic to store the path to our installation of <i>gkno</i> so we don't need to keep typing the whole path.
  </div>
  <div class="panel panel-danger">
    <ul class="code-list">
      <li>mkdir bam-demo</li>
      <li>GKNO=&lt;gkno_launcher path&gt;</li>
      <li>cp $GKNO/resources/tutorial/current/mutated_genome.bam bam-demo</li>
      <li>cp $GKNO/resources/tutorial/current/chr20_fragment.fa bam-demo</li>
      <li>cd bam-demo</li>
    </ul>
  </div>
  <div class='description'>
    A command <span class="code-list code-span panel panel-danger">NAME=VALUE</span> means that we have created an environment variable with the name <span class="code-list code-span panel panel-danger">NAME</span> which contains the value <span class="code-list code-span panel panel-danger">VALUE</span>. Whenever we use the variable, we need to prepend <span class="code-list code-span panel panel-danger">$</span> to it and it will assume the value we gave to it. From now on, every time we need to access the directory where <i>gkno</i> is installed, we will just use the environment variable <span class="code-list code-span panel panel-danger">$GKNO</span>. User-defined environment variables like this only exist for the duration of the session. If you log out and log in again, the environment variable will no longer exist. There are ways to make them permanent, but that is beyond the scope of this tutorial.
  <div class='description'>
    Next, we will use the bash 'alias' command to define some simple commands on the command line. This command lets us assign an operation to a command of our choice. In this case, we will link the command <i>samtools</i> to the <i>samtools</i> tool executable and, similarly, use the <i>bamtools</i> command to link to the <i>bamtools</i> tool executable.
  </div>
  <div class="panel panel-danger">
    <ul class="code-list">
      <li>alias samtools="$GKNO/tools/samtools/samtools"</li>
      <li>alias bamtools="$GKNO/tools/bamtools/bin/bamtools"</li>
    </ul>
  </div>

  <!-- tools -->
  <div class='section-subtitle'>bamtools and samtools</div>
  <div class='description'>
    We will be using the tools <i>bamtools</i> and <i>samtools</i> as part of this tutorial. These tools are common tools for working with BAM files and have a lot of overlap. For most basic tasks, you can use either one of these tools. There are certain functions that can only be performed by one or the other tools, so we will make ourselves familiar with both. There are still other tools that perform additional tasks that we are not looking at here.
  </div>

  <!-- samtools help -->
  <div class='section-subtitle'>samtools help</div>
  <div class='description'>
    To see help for samtools, type:
  </div>
  <div class="panel panel-danger">
    <ul class="code-list">
      <li>samtools</li>
    </ul>
  </div>
  <div class='description'>
    You can see that there are a lot of subcommands for <i>samtools</i>. Each of these performs a different task, some of which we will go through in this tutorial. To get help on individual subcommands, simply include the subcommand you want to know about.
  </div>
  <div class="panel panel-danger">
    <ul class="code-list">
      <li>samtools view</li>
      <li>samtools sort</li>
    </ul>
  </div>
  <div class='description'>
    The subcommand help gives you the format of the command line expected by <i>samtools</i>, including the input and output files as well as a list of all of the additional arguments that can be selected along with a brief description of their role. We can see more detailed help with the following command:
  </div>
  <div class="panel panel-danger">
    <ul class="code-list">
      <li>samtools view -?</li>
    </ul>
  </div>

  <!-- bamtools help -->
  <div class='section-subtitle'>bamtools help</div>
  <div class='description'>
    Similar to <i>samtools</i>, to see the available commands for <i>bamtools</i>, type:
  </div>
  <div class="panel panel-danger">
    <ul class="code-list">
      <li>bamtools</li>
    </ul>
  </div>
  <div class='description'>
    To see help for the <i>bamtools</i> subcommands, type:
  </div>
  <div class="panel panel-danger">
    <ul class="code-list">
      <li>bamtools convert help</li>
      <li>bamtools count help</li>
    </ul>
  </div>
  <div class='description'>
    Again, like <i>samtools</i>, a description of the tool and the arguments that can be supplied is given.
  </div>

  <!-- samtools view -->
  <div class='section-subtitle'>samtools view</div>
  <div class='description'>
    The first thing that we probably want to do with the BAM file is look at it. Since the BAM file is a compressed binary format, using standard bash commands like 'more' or 'less', won't work.
  </div>
  <div class="panel panel-danger">
    <ul class="code-list">
      <li>samtools view sample.bam | less -S</li>
    </ul>
  </div>
  <div class='description'>
    The <span class="code-list code-span panel panel-danger">|</span> character 'pipes' the output from the <i>samtools</i> command into the bash command. The command <span class="code-list code-span panel panel-danger">less -S</span> prints the results of the <i>samtools</i> command to screen and makes sure that the lines don't 'wrap' around. Try the same command again without the <span class="code-list code-span panel panel-danger">-S</span> and you'll see what this looks like without wrapping. Note: press 'q' to quit out of the view of the BAM file.
  </div>

  <!-- BAM header -->
  <div class='section-subtitle'>The BAM header</div>
  <div class='description'>
    You probably noticed that we didn't see the BAM header. It is there, we just need to tell <i>samtools</i> that we want to see it.
  </div>
  <div class="panel panel-danger">
    <ul class="code-list">
      <li>samtools view -H sample.bam | less -S</li>
    </ul>
  </div>
  <div class='description'>
    Including <span class="code-list code-span panel panel-danger">-H</span> now tells <i>samtools</i> to only view the header and so this is all we see. If we instead use <span class="code-list code-span panel panel-danger">-h</span>, we will see the whole BAM file, header, alignments and all.
  </div>
  <div class="panel panel-danger">
    <ul class="code-list">
      <li>samtools view -h sample.bam | less -S</li>
    </ul>
  </div>

  <!-- samtools sort -->
  <div class='section-subtitle'>samtools sort</div>
  <div class='description'>
    You may have also noticed that the BAM file is not sorted. Aligners typically take FASTQ files and try to align them to the reference sequence. The resulting BAM file contains all of the alignments in the same order that they appeared in the FASTQ file, which is completely random in terms of genomic coordinates. For most downstream tools (e.g. <a href="http://iobio.io"><i>iobio</i></a>, <i>IGV</i> or variant callers), the BAM file needs to be sorted by genomic coordinates. We can easily do this:
  </div>
  <div class="panel panel-danger">
    <ul class="code-list">
      <li>samtools sort sample.bam sample_sorted</li>
    </ul>
  </div>
  <div class='description'>
    Note that <i>samtools</i> only wants a prefix for the output file. If you set the output file as <i>sample_sorted.bam</i>, you will have generated a file called <i>sample_sorted.bam.bam</i>!! Now let's take a look at the new BAM file and check that it's sorted properly.
  </div>
  <div class="panel panel-danger">
    <ul class="code-list">
      <li>samtools view -h sample_sorted.bam | less -S</li>
    </ul>
  </div>
  <div class='description'>
    You can see that the alignments are in coordinate order, but you can also see that the first line in the BAM header states (with the <span class="code-list code-span panel panel-danger">SO</span> tag in the <span class="code-list code-span panel panel-danger">@HD</span> line, that the Sort Order is 'coordinate' as required.
  </div>

  <!-- samtools index -->
  <div class='section-subtitle'>samtools index</div>
  <div class='description'>
    Now we have a sorted BAM file, but this usually isn't enough for most tools. For example, both <a href="http://iobio.io"><i>iobio</i></a> and <i>IGV</i> want to look at specific regions within the BAM file and don't want to have to search through the entire file to find the region, since the search would take an extremely long time. To get around this, BAM files need to be indexed. This index provides tools a map of the data in the BAM file, so that they can quickly jump to the regions of interest. So, let's index our BAM file:
  </div>
  <div class="panel panel-danger">
    <ul class="code-list">
      <li>samtools index sample_sorted.bam</li>
    </ul>
  </div>
  <div class='description'>
    If we look in our directory (<span class="code-list code-span panel panel-danger">ls</span>), we will see the index file with the extension '.bai'. Let's exploit this index to see how many reads fall into the region '20000-30000' on chromosome 20.
  </div>
  <div class="panel panel-danger">
    <ul class="code-list">
      <li>samtools view sample_sorted.bam 20:20000-30000 | less -S</li>
    </ul>
  </div>
  <div class='description'>
    What if we want to know how many reads fall into this region?
  </div>
  <div class="panel panel-danger">
    <ul class="code-list">
      <li>samtools view sample_sorted.bam 20:20000-30000 | wc -l</li>
    </ul>
  </div>

  <!-- BAM flag -->
  <div class='section-subtitle'>The BAM flag</div>
  <div class='description'>
    The BAM flag (the second column) encodes lots of useful information, but isn't particularly readable by humans. We can use <i>samtools</i> to convert the bitwise flag into a list of characters. This is more readable than the simple number, but is still not great. A better way of viewing the meaning of the flag is <a href="https://broadinstitute.github.io/picard/explain-flags.html">here</a>, where you can type in the number and see a full explanation of the flag.
  </div>
  <div class='description'>
    What if we want to look at all the reads in the BAM file that are marked as being in a pair. We can use the following command to do this:
  </div>
  <div class="panel panel-danger">
    <ul class="code-list">
      <li>samtools view -f 0x1 sample_sorted.bam | less -S</li>
    </ul>
  </div>
  <div class='description'>
    You will see that the most of the reads have flags '99' or '163'. If we look at the flag explanations <a href="https://broadinstitute.github.io/picard/explain-flags.html">here</a>, we can see that both of these flags include the 1 (or 0x1) bit referring to the reads being paired. While this is functional, it still isn't that easy to use. This is somewhere where we can use <i>bamtools</i> more effectively. We can generate the same list of reads with the command:
  </div>
  <div class="panel panel-danger">
    <ul class="code-list">
      <li>bamtools filter -isPaired true -in sample_sorted.bam | samtools view - | less -S</li>
    </ul>
  </div>
  <div class='description'>
    There are a few differences between the <i>samtools</i> and the <i>bamtools</i> command lines. Firstly, <i>bamtools</i> requires that the input BAM file is identified with the <span class="code-list code-span panel panel-danger">-in</span> argument. Secondly, by default, <i>samtools view</i> outputs a uncompressed SAM file, by default, where <i>bamtools</i> will only output a SAM file when instructed to do so with the <span class="code-list code-span panel panel-danger">bamtools convert -format sam</span> command. So in the above command, we pipe the output of the <i>bamtools</i> command to <i>samtools</i> for viewing. We can check that these two commands are equivalent by counting the number of reads:
  </div>
  <div class="panel panel-danger">
    <ul class="code-list">
      <li>samtools view -f 0x1 sample_sorted.bam | wc -l</li>
      <li>bamtools filter -isPaired true -in sample_sorted.bam | samtools view - | wc -l</li>
    </ul>
  </div>
  <div class='description'>
    This is much easier to see what we are filtering on. The command <span class="code-list code-span panel panel-danger">bamtools filter help</span> gives a list of all the different filters that can be applied, which are more readable for people!
  </div>

  <!-- samtools tview -->
  <div class='section-subtitle'>samtools tview</div>
  <div class='description'>
    <i>IGV</i> is a very useful tool for looking at alignments together, but sometimes you just want a quick look at the data in a particular region. For example, we might want to see all the reads spanning a particular exon. Let's just quickly take a look at the reads that fall into the range 71,800 - 71,900 on chromosome 20.
  </div>
  <div class="panel panel-danger">
    <ul class="code-list">
      <li>samtools tview -p 20:71800-71900 sample_sorted.bam chr20_fragment.fa</li>
    </ul>
  </div>
  <div class='description'>
    We can clearly see an insertion of the sequence 'ACCAGT' in most of these reads. Lower case letters (or ',') represent alignment to the reverse strand, whereas uppercase letters (and '.') represent alignment to the forward strand. Press <span class="code-list code-span panel panel-danger">?</span> to view options for changing the view in <i>tview</i>, for example changing the colour scheme, or showing all bases, not just those that differ from the reference.
  </div>
</div>
